contracts/proposal/cwd-proposal-delegate/src/contract.rs

```diff
@@ -0,0 +1,238 @@
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Delegate {
+            delegate,
+            msgs,
+            expiration,
+            policy_irrevocable,
+            policy_preserve_on_failure,
+        } => {
+            let policy_irrevocable = policy_irrevocable.unwrap_or(DEFAULT_POLICY_IRREVOCABLE);
+            let policy_preserve_on_failure =
+                policy_preserve_on_failure.unwrap_or(DEFAULT_POLICY_PRESERVE_ON_FAILURE);
+            let delegate = deps.api.addr_validate(&delegate)?;
+            execute_delegate(
+                deps,
+                env,
+                info,
+                Delegation {
+                    delegate,
+                    msgs,
+                    expiration,
+                    policy_irrevocable,
+                    policy_preserve_on_failure,
+                },
+            )
+        }
+        ExecuteMsg::RemoveDelegation { delegation_id } => {
+            execute_remove_delegation(deps, env, info, delegation_id)
+        }
+        ExecuteMsg::Execute { delegation_id } => execute_execute(deps, env, info, delegation_id),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ContractError> {
+    match msg.id {
+        REPLY_ID_EXECUTE_PROPOSAL_HOOK => {
+            let id = EXECUTE_CTX.load(deps.storage)?;
+            let Delegation {
+                policy_preserve_on_failure,
+                ..
+            } = DELEGATIONS.load(deps.storage, id)?;
+
+            if let SubMsgResult::Err(_) = msg.result {
+                // && with `let` expression above gives unstable Rust error
+                if policy_preserve_on_failure {
+                    return Ok(Response::default()
+                        .add_attribute("execute_failed_but_preserved", id.to_string()));
+                }
+            }
+            // Delete delegation in both success and error case
+            DELEGATIONS.remove(deps.storage, id);
+            Ok(Response::default())
+        }
+        _ => Err(ContractError::Std(StdError::GenericErr {
+            msg: "Reply handler ID not found".into(),
+        })),
+    }
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::DelegationCount {} => Ok(to_binary(&query_delegation_count(deps)?)?),
+        QueryMsg::Delegation { delegation_id } => {
+            Ok(to_binary(&query_delegation(deps, delegation_id)?)?)
+        }
+    }
+}
+
+// MARK: Query helpers
+
+fn query_delegation_count(deps: Deps) -> StdResult<DelegationCountResponse> {
+    let count = DELEGATION_COUNT.load(deps.storage)?;
+    Ok(DelegationCountResponse { count })
+}
+
+fn query_delegation(deps: Deps, delegation_id: u64) -> StdResult<DelegationResponse> {
+    let delegation = DELEGATIONS.load(deps.storage, delegation_id)?;
+    Ok(delegation)
+}
```

Comment:
probably want to make this a paginated function so that FE can list all of the delegations pending. `packages/cw-paginate` should make that simple

Reply:
cool!